src/coreclr/jit/importer.cpp

```diff
@@ -3764,6 +3764,65 @@ GenTree* Compiler::impInitClass(CORINFO_RESOLVED_TOKEN* pResolvedToken)
     return node;
 }
 
+//------------------------------------------------------------------------
+// impGetNodeFromLocal: Tries to return the node that's assigned
+// to the provided local.
+//
+// Arguments:
+//    node - GT_LCL_VAR whose value is searched for
+//
+// Return Value:
+//    The tree representing the node assigned to the variable when possible,
+//    nullptr otherwise.
+//
+GenTree* Compiler::impGetNodeFromLocal(GenTree* node)
+{
+    assert(node != nullptr);
+    assert(node->OperIs(GT_LCL_VAR));
+
+    unsigned lclNum = node->AsLclVarCommon()->GetLclNum();
+
+    if (lvaTable[lclNum].lvSingleDef == 0)
+    {
+        return nullptr;
+    }
+
+    auto findValue = [&](Statement* stmtList) -> GenTree* {
+        for (Statement* stmt : StatementList(stmtList))
+        {
+            GenTree* root = stmt->GetRootNode();
+            if (root->OperIs(GT_ASG) && root->AsOp()->gtOp1->OperIs(GT_LCL_VAR) &&
+                root->AsOp()->gtOp1->AsLclVarCommon()->GetLclNum() == lclNum)
+            {
+                return root->AsOp()->gtOp2;
+            }
+        }
+        return nullptr;
+    };
+
+    GenTree*    valueNode = findValue(impStmtList);
+    BasicBlock* bb        = fgFirstBB;
```

Comment:
I'm still ambivalent about this non-local searching.

Seems like we could have cases where we are able to do this transformation in Tier0-instr but not in a subsequent OSR, because OSR doesn't have the same `fgFirstBB`.

I suppose one fix is not to do this transformation in Tier0-instr (though removing the need to have a method probe is appealing)... but only if we can then always do the same in a subsequent rejit.